### ISSUE TITLE
tablesorter.filterFormatter.select2 doesn't work with special characters

### ISSUE THREAD

**Issue by gurditta** [Mottie/tablesorter#796]
Code for the Filter : 

```js
3 : function($cell, indx){
	return $.tablesorter.filterFormatter.select2( $cell, indx, {
		match : false,
		placeholder: 'search...'
	});
},
```

Issue is in the attached screenshot
![image](https://cloud.githubusercontent.com/assets/3133431/5784985/31dfee26-9d97-11e4-8720-cd1acbfbd4f6.png)

I tried making  `match : true,` which also doesn't work

**Comment by Mottie** [comment]
Hi @gurditta!

It looks like the main issue is that the filterFormatter was written for select2 v3.4.6. When the value contains a comma (e.g. `A & A Recycling, LLC.`) it splits it into an array; so the value of that select2 becomes `["A & A", "LLC."]` ([demo](http://jsfiddle.net/Mottie/e6b5rq4d/)).

It looks like the issue was fixed in the latest version of select2, v4.0.0 beta 2, where the result is the proper string `["A & A, LLC."]` ([demo](http://jsfiddle.net/Mottie/wphqwvLf/)).

So what happens is that when the select2 filterformatter code sees an array, it thinks you are looking for two separate tags; so an exact match won't find `"A & A"` and a match will find the "A & A" but it won't match any select2 tags and thus the correct results show up, but no tags are visible in the select2 input.

Basically this can be solved in two ways:
1. Remove the commas
2. Update filterformatter for select2 to work with select2 v4.0.0

I don't have that much free time, so option 2 probably won't be happening any time soon. And since I plan on releasing a new filter option called `filter_customVal` which will pretty much make the `filterFormatter` option obsolete, that makes an update to the select2 filterformatter even less likely.

I'm not saying I won't ever update the select2 filterformatter. I just don't have the time right now.

**Comment by gurditta** [comment]
Hi @Mottie 
Thanks much for replying, really appreciate that.
But the problem still persists (another problem came up) a weird behavior found, Please have a keen look at the below screenshot i've tried to make it as understandable as i could :)

![image](https://cloud.githubusercontent.com/assets/3133431/5786874/c8f43f86-9db2-11e4-9057-96b69f6b01f2.png)

It's not only about the comma you're talking about.

**Comment by gurditta** [comment]
/*! Filter widget select2 formatter function - updated 7/17/2014 (v2.17.5)
 * requires: jQuery 1.7.2+, tableSorter (FORK) 2.16+, filter widget 2.16+ and select2 v3.4.6+ plugin
 */
/*jshint browser:true, jquery:true, unused:false */
/*global jQuery: false */
;(function($){
"use strict";

var ts = $.tablesorter || {};
ts.filterFormatter = ts.filterFormatter || {};

/************************\
 Select2 Filter Formatter
\************************/
ts.filterFormatter.select2 = function($cell, indx, select2Def) {
	var o = $.extend({
		// select2 filter formatter options
		cellText : '', // Text (wrapped in a label element)
		match : true, // adds "filter-match" to header
		value : '',
		// include ANY select2 options below
		multiple : true,
		width : '100%'

	}, select2Def ),
	arry, data,
	c = $cell.closest('table')[0].config,
	wo = c.widgetOptions,
	// Add a hidden input to hold the range values
	$input = $('<input class="filter" type="hidden">')
	.appendTo($cell)
	// hidden filter update namespace trigger by filter widget
	.bind('change' + c.namespace + 'filter', function(){
		var val = this.value;
		val = val.replace(/[/()$^]/g, '').split('|');
		$cell.find('.select2').select2('val', val);
		updateSelect2();
	}),
	$header = c.$headers.filter('[data-column="' + indx + '"]:last'),
	onlyAvail = $header.hasClass(wo.filter_onlyAvail),
	$shcell = [],
	matchPrefix = o.match ? '' : '^',
	matchSuffix = o.match ? '' : '$',

	// this function updates the hidden input and adds the current values to the header cell text
	updateSelect2 = function() {
		var v = $cell.find('.select2').select2('val') || o.value || '';
		$input
		// add regex, so we filter exact numbers
		.val( $.isArray(v) && v.length && v.join('') !== '' ? '/(' + matchPrefix + (v || []).join(matchSuffix + '|' + matchPrefix) + matchSuffix + ')/' : '' )
		.trigger('search').end()
		.find('.select2').select2('val', v);
		// update sticky header cell
		if ($shcell.length) {
			$shcell
			.find('.select2').select2('val', v);
		}
	},

	// get options from table cell content or filter_selectSource (v2.16)
	updateOptions = function(){
		data = [];
		arry = ts.filter.getOptionSource(c.$table[0], indx, onlyAvail) || [];
		// build select2 data option
		$.each(arry, function(i,v){
			data.push({id: v, text: v});
		});
		o.data = data;
	};

	// get filter-match class from option
	$header.toggleClass('filter-match', o.match);
	if (o.cellText) {
		$cell.prepend('<label>' + o.cellText + '</label>');
	}

	// don't add default in table options if either ajax or
	// data options are already defined
	if (!(o.ajax && !$.isEmptyObject(o.ajax)) && !o.data) {
		updateOptions();
		if (onlyAvail) {
			c.$table.bind('filterEnd', function(){
				updateOptions();
				$cell.add($shcell).find('.select2').select2(o);
			});
		}
	}

	// add a select2 hidden input!
	$('<input class="select2 select2-' + indx + '" type="hidden" />')
	.val(o.value)
	.appendTo($cell)
	.select2(o)
	.bind('change', function(){
		updateSelect2();
	});

	// update select2 from filter hidden input, in case of saved filters
	c.$table.bind('filterFomatterUpdate', function(){
		// value = '/(^x$|^y$)/' => 'x,y'
		var val = c.$table.data('lastSearch')[indx] || '';
		val = val.replace(/[/()$^]/g, '').split('|');
		$cell.find('.select2').select2('val', val);
		updateSelect2();
		ts.filter.formatterUpdated($cell, indx);
	});

	// has sticky headers?
	c.$table.bind('stickyHeadersInit', function(){
		$shcell = c.widgetOptions.$sticky.find('.tablesorter-filter-row').children().eq(indx).empty();
		// add a select2!
		$('<input class="select2 select2-' + indx + '" type="hidden">')
		.val(o.value)
		.appendTo($shcell)
		.select2(o)
		.bind('change', function(){
			$cell.find('.select2').select2('val', $shcell.find('.select2').select2('val') );
			updateSelect2();
		});
		if (o.cellText) {
			$shcell.prepend('<label>' + o.cellText + '</label>');
		}

	});

	// on reset
	c.$table.bind('filterReset', function(){
		$cell.find('.select2').select2('val', o.value || '');
		setTimeout(function(){
			updateSelect2();
		}, 0);
	});

	updateSelect2();
	return $input;
};

})(jQuery);

**Comment by Mottie** [comment]
Ok, there was a bug in the select2 filterFormatter code in that it was not escaping special regex characters - like the parenthesis!

So I've fixed that part of the select2 filterFormatter code ([ref](https://github.com/Mottie/tablesorter/commit/66b5682cf021892b8f16d1f441ec079d73664b28)) - here is [a demo](http://jsfiddle.net/Mottie/abkNM/5040/); please note that the commas still need to be removed!

You can get a copy of the fixed code from here : https://github.com/Mottie/tablesorter/blob/working/js/jquery.tablesorter.widgets-filter-formatter-select2.js

**Comment by gurditta** [comment]
Thanks a lot for the fix.

For the comma problem, i'm not able to understand whether I'll have to remove the comma from the table data or the drop down

**Comment by Mottie** [comment]
Hmm, I actually had to dig into the code and found that you can change the `separator` setting ([demo](http://jsfiddle.net/Mottie/abkNM/5041/))

```js
0: function ($cell, indx) {
    return $.tablesorter.filterFormatter.select2($cell, indx, {
        separator: "|",
        match: true // adds "filter-match" to header
    });
},
```

Gotta love incomplete documentation.

**Comment by Mottie** [comment]
By the way, you can save yourself a lot of headache/work by adding a class name to your headers:

```html
<th class="select2">AlphaNumeric</th>
<th class="select2">AlphaNumeric 2</th>
<th>Numeric</th>
<th class="select2">Animals</th>
<th>Sites</th>
```

then instead of defining the `filterFormatter` by column index, just use the class name ([demo](http://jsfiddle.net/Mottie/abkNM/5043/)):

```js
filter_formatter: {
	// target all select2 columns
	'.select2' : function ($cell, indx) {
		return $.tablesorter.filterFormatter.select2($cell, indx, {
			tags: true,
			tokenSeparators: [',', ' ', '|'],
			//separator: "|",
			match: true // adds "filter-match" to header
		});
	}
}
```

**Comment by gurditta** [comment]
That's cool @Mottie 
It's working fine now
Thanks much!!!
